src/app/icd/client/ICDRefreshKeyInfo.cpp

```diff
@@ -0,0 +1,128 @@
+ *    limitations under the License.
+ */
+
+#include "ICDRefreshKeyInfo.h"
+#include "CheckInDelegate.h"
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app/CommandPathParams.h>
+#include <app/InteractionModelEngine.h>
+#include <app/InteractionModelTimeout.h>
+#include <app/OperationalSessionSetup.h>
+
+namespace chip {
+namespace app {
+
+ICDRefreshKeyInfo::ICDRefreshKeyInfo(CheckInDelegate * apCheckInDelegate, const ICDClientInfo aICDClientInfo,
+                                     const ICDClientStorage * aICDClientStorage) :
+    mOnConnectedCallback(HandleDeviceConnected, this),
+    mOnConnectionFailureCallback(HandleDeviceConnectionFailure, this)
+{
+    mpCheckInDelegate  = apCheckInDelegate;
+    mICDClientInfo     = aICDClientInfo;
+    mpICDClientStorage = const_cast<ICDClientStorage *>(aICDClientStorage);
+}
+
+void ICDRefreshKeyInfo::OnResponse(chip::app::CommandSender * apCommandSender, const chip::app::ConcreteCommandPath & aPath,
+                                   const chip::app::StatusIB & aStatus, chip::TLV::TLVReader * aData)
+{
+    app::Clusters::IcdManagement::Commands::RegisterClientResponse::DecodableType responseData;
+    mError = DataModel::Decode(*aData, responseData);
+    if (mError != CHIP_NO_ERROR)
+    {
+        ChipLogError(ICD, "Failed to decode re-registration response data: %" CHIP_ERROR_FORMAT, mError.Format());
+        return;
+    }
+    // Update the ICDClientInfo with new key and start counter and store it to persistence
+    mICDClientInfo.start_icd_counter = responseData.ICDCounter;
+    mError                           = mpICDClientStorage->SetKey(mICDClientInfo, mNewKey.Span());
+    if (mError != CHIP_NO_ERROR)
+    {
+        ChipLogError(ICD, "Failed to store the new key after re-registration: %" CHIP_ERROR_FORMAT, mError.Format());
+    }
+}
+void ICDRefreshKeyInfo::OnError(const chip::app::CommandSender * apCommandSender, CHIP_ERROR aError)
+{
+    mError = aError;
+}
+
+void ICDRefreshKeyInfo::OnDone(chip::app::CommandSender * apCommandSender)
+{
+    if (mError == CHIP_NO_ERROR)
+    {
+        mpCheckInDelegate->OnRegistrationUpdateComplete(mICDClientInfo, this);
+        return;
+    }
+    mpCheckInDelegate->OnRegistrationUpdateFailure(this);
+}
+
+CHIP_ERROR ICDRefreshKeyInfo::RegisterClientWithNewKey(Messaging::ExchangeManager & exchangeMgr,
+                                                       const SessionHandle & sessionHandle)
+{
+    using namespace Clusters::IcdManagement;
+
+    mRegisterCommandSender.SetExchangeManager(&exchangeMgr);
+    mRegisterCommandSender.SetCommandSenderDelegate(this);
+
+    auto commandPathParams = CommandPathParams(0, 0, Id, Commands::RegisterClient::Id, (CommandPathFlags::kEndpointIdValid));
+
+    ReturnErrorOnFailure(mRegisterCommandSender.PrepareCommand(commandPathParams));
+
+    chip::TLV::TLVWriter * writer = mRegisterCommandSender.GetCommandDataIBTLVWriter();
```

Comment:
Please, no.  Use the type-safe methods on CommandSender for sending a Commands::RegisterClient::Type, which is what you should create and fill in here, instead of hand-writing the TLV bits.

Reply:
Okay. Will use the type-safe methods.